// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_outer_classname = "PostsProto";
option objc_class_prefix = "PTS";

package posts;

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}

  rpc GetPost (GetRequest) returns (GetReply) {}

  rpc AddPost (AddRequest) returns (AddReply) {}

  rpc DeletePost (DeleteRequest) returns (DeleteReply) {}

  rpc UpdatePost (UpdateRequest) returns (UpdateReply) {}

  rpc ListPost (ListRequest) returns (ListReply) {}

  rpc GetStatistics (StatisticRequest) returns (StatisticResponse);

  rpc TopPosts () returns(TopPostResponse);

  rpc TopUsers () returns(TopUsersResponse);
}


message Post {
  int32 id = 1;
  string login = 2;
  string description = 3;
  bytes image = 4;
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}


message GetRequest {
  int32 id = 1;
  string login = 2;
}

message GetReply {
  bool success = 1;
  string message = 2;
  Post post = 3;
}


message AddRequest {
  Post post = 1;
}

message AddReply {
  bool success = 1;
  string message = 2;
}


message DeleteRequest {
  int32 id = 1;
  string login = 2;
}

message DeleteReply {
  bool success = 1;
  string message = 2;
}


message UpdateRequest {
  Post post = 2;
}

message UpdateReply {
  bool success = 1;
  string message = 2;
}


message ListRequest {
  string login = 1;
}

message ListReply {
  bool success = 1;
  string message = 2;
  repeated Post list = 3;
}
